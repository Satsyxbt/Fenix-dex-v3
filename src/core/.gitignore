node_modules
.env
cache
artifacts
typechain
corpus
.idea
bytecode
hashCalc.js
AllBlocks_timestamped.json
results.json