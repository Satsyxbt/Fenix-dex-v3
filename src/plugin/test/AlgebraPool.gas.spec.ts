import { ethers } from 'hardhat';
import { Wallet } from 'ethers';
import { loadFixture, reset as resetNetwork } from '@nomicfoundation/hardhat-network-helpers';
import { MockTimeAlgebraPool } from '../../core/typechain';
import { MockTimeAlgebraBasePluginV1, MockTimeDSFactory, MockTimeVirtualPool } from '../typechain';
import { expect } from './shared/expect';

import { algebraPoolDeployerMockFixture } from './shared/externalFixtures';

import snapshotGasCost from './shared/snapshotGasCost';

import {
  expandTo18Decimals,
  FeeAmount,
  getMinTick,
  encodePriceSqrt,
  TICK_SPACINGS,
  createPoolFunctions,
  SwapFunction,
  MintFunction,
  getMaxTick,
  MaxUint128,
  SwapToPriceFunction,
} from '../../core/test/shared/utilities';
import { ZERO_ADDRESS } from './shared/fixtures';

describe('AlgebraPool gas tests [ @skip-on-coverage ]', () => {
  let wallet: Wallet, other: Wallet;

  before('create fixture loader', async () => {
    await resetNetwork();
    [wallet, other] = await (ethers as any).getSigners();
  });

  const startingPrice = encodePriceSqrt(100001, 100000);
  const startingTick = 0;
  const tickSpacing = TICK_SPACINGS[FeeAmount.MEDIUM];
  const minTick = getMinTick(tickSpacing);
  const maxTick = getMaxTick(tickSpacing);

  async function gasTestFixture() {
    const fix = await algebraPoolDeployerMockFixture();
    const pool = await fix.createPool();

    const mockPluginFactoryFactory = await ethers.getContractFactory('MockTimeDSFactory');
    const mockPluginFactory = (await mockPluginFactoryFactory.deploy(fix.factory)) as any as MockTimeDSFactory;

    await mockPluginFactory.createPlugin(pool, ZERO_ADDRESS, ZERO_ADDRESS);
    const pluginAddress = await mockPluginFactory.pluginByPool(pool);

    const mockDSOperatorFactory = await ethers.getContractFactory('MockTimeAlgebraBasePluginV1');
    const plugin = mockDSOperatorFactory.attach(pluginAddress) as any as MockTimeAlgebraBasePluginV1;

    await pool.setPlugin(plugin);

    const advanceTime = async (secs: any) => {
      await pool.advanceTime(secs);
      await plugin.advanceTime(secs);
    };

    const { swapExact0For1, swapExact1For0, swapToHigherPrice, mint, swapToLowerPrice } = createPoolFunctions({
      swapTarget: fix.swapTargetCallee,
      token0: fix.token0,
      token1: fix.token1,
      pool,
    });

    const virtualPoolMockFactory = await ethers.getContractFactory('MockTimeVirtualPool');
    const virtualPoolMock = (await virtualPoolMockFactory.deploy()) as any as MockTimeVirtualPool;

    await pool.initialize(encodePriceSqrt(1, 1));
    await pool.setCommunityVault(wallet.address);

    await advanceTime(1);
    await mint(wallet.address, minTick, maxTick, expandTo18Decimals(2));
    await swapExact0For1(expandTo18Decimals(1), wallet.address);
    await advanceTime(1);
    await swapToHigherPrice(startingPrice, wallet.address);
    await advanceTime(1);
    expect((await pool.globalState()).tick).to.eq(startingTick);
    expect((await pool.globalState()).price).to.eq(startingPrice);

    return {
      advanceTime,
      pool,
      plugin,
      virtualPoolMock,
      mockPluginFactory,
      swapExact0For1,
      swapExact1For0,
      mint,
      swapToHigherPrice,
      swapToLowerPrice,
    };
  }

  let swapExact0For1: SwapFunction;
  let swapExact1For0: SwapFunction;
  let swapToHigherPrice: SwapToPriceFunction;
  let pool: MockTimeAlgebraPool;
  let plugin: MockTimeAlgebraBasePluginV1;
  let virtualPoolMock: MockTimeVirtualPool;
  let mockPluginFactory: MockTimeDSFactory;
  let mint: MintFunction;
  let advanceTime: any;

  for (const communityFee of [0, 60]) {
    describe(communityFee > 0 ? 'fee is on' : 'fee is off', () => {
      const gasTestCommunityFeeFixture = async () => {
        const fix = await gasTestFixture();
        if (communityFee > 0) {
          await fix.pool.setCommunityFee(communityFee);
          await fix.swapExact0For1(expandTo18Decimals(1), wallet.address);
          await fix.swapToHigherPrice(startingPrice, wallet.address);
          await fix.advanceTime(1);
        }
        await fix.advanceTime(1);
        await fix.swapExact0For1(1, wallet.address);
        await fix.advanceTime(1);
        expect((await fix.pool.globalState()).tick).to.eq(startingTick);
        return fix;
      };

      beforeEach('load the fixture', async () => {
        ({ advanceTime, swapExact0For1, swapExact1For0, pool, plugin, virtualPoolMock, mockPluginFactory, mint, swapToHigherPrice } =
          await loadFixture(gasTestCommunityFeeFixture));
      });

      describe('#swap', async () => {
        describe('#swapExact1For0', () => {
          it('first swap in block with no tick movement', async () => {
            await snapshotGasCost(swapExact1For0(2000, wallet.address));
            expect((await pool.globalState()).price).to.not.eq(startingPrice);
            expect((await pool.globalState()).tick).to.eq(startingTick);
          });

          it('first swap in block moves tick, no initialized crossings', async () => {
            await snapshotGasCost(swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick + 1);
          });

          it('second swap in block with no tick movement', async () => {
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            expect((await pool.globalState()).tick).to.eq(startingTick + 1);
            await snapshotGasCost(swapExact1For0(2000, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick + 1);
          });
        });

        describe('#swapExact0For1', () => {
          it('first swap in block with no tick movement', async () => {
            await snapshotGasCost(swapExact0For1(2000, wallet.address));
            expect((await pool.globalState()).price).to.not.eq(startingPrice);
            expect((await pool.globalState()).tick).to.eq(startingTick);
          });

          it('first swap in block with no tick movement, static fee', async () => {
            await plugin.changeFeeConfiguration({
              alpha1: 0,
              alpha2: 0,
              beta1: 0,
              beta2: 0,
              gamma1: 1,
              gamma2: 1,
              baseFee: 100,
            });
            await snapshotGasCost(swapExact0For1(2000, wallet.address));
            expect((await pool.globalState()).price).to.not.eq(startingPrice);
            expect((await pool.globalState()).tick).to.eq(startingTick);
          });

          it('first swap in block moves tick, no initialized crossings', async () => {
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick - 1);
          });

          it('second swap in block with no tick movement', async () => {
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            expect((await pool.globalState()).tick).to.eq(startingTick - 1);
            await snapshotGasCost(swapExact0For1(2000, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick - 1);
          });

          it('second swap in block moves tick, no initialized crossings', async () => {
            await swapExact0For1(1000, wallet.address);
            expect((await pool.globalState()).tick).to.eq(startingTick);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick - 1);
          });

          it('first swap in block, large swap, no initialized crossings', async () => {
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(10), wallet.address));
            expect((await pool.globalState()).tick).to.eq(-35586);
          });

          it('first swap in block, large swap crossing several initialized ticks', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            expect((await pool.globalState()).tick).to.eq(startingTick);
            await advanceTime(15);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing); // we crossed the last tick
          });

          it('several large swaps with pauses', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            expect((await pool.globalState()).tick).to.eq(startingTick);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60 * 60);
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60 * 60);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(5 * 60 * 60);
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(19 * 60 * 60);
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing); // we crossed the last tick
          });

          it('small swap after several large swaps with pauses', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            expect((await pool.globalState()).tick).to.eq(startingTick);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60 * 60);
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60 * 60);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(5 * 60 * 60);
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(19 * 60 * 60);
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await advanceTime(60);
            await snapshotGasCost(swapExact0For1(1000, wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(0);
          });

          it('first swap in block, large swap crossing a single initialized tick', async () => {
            await mint(wallet.address, minTick, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            await advanceTime(15);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 2 * tickSpacing); // we crossed the last tick
          });

          it('second swap in block, large swap crossing several initialized ticks', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing);
          });

          it('second swap in block, large swap crossing a single initialized tick', async () => {
            await mint(wallet.address, minTick, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            await swapExact0For1(expandTo18Decimals(1) / 10000n, wallet.address);
            expect((await pool.globalState()).tick).to.be.gt(startingTick - 2 * tickSpacing); // we didn't cross the initialized tick
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 2 * tickSpacing); // we crossed the last tick
          });

          it('large swap crossing several initialized ticks after some time passes', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            await swapExact0For1(2, wallet.address);
            await advanceTime(1);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing);
          });

          it('large swap crossing several initialized ticks second time after some time passes', async () => {
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            await swapExact0For1(expandTo18Decimals(1), wallet.address);
            await swapToHigherPrice(startingPrice, wallet.address);
            await advanceTime(1);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(tickSpacing * -4);
          });
        });

        describe('farming connected', async () => {
          beforeEach('connect virtual pool', async () => {
            await mockPluginFactory.setFarmingAddress(wallet);
            await plugin.connect(wallet).setIncentive(virtualPoolMock);
          });

          it('first swap in block with no tick movement', async () => {
            await snapshotGasCost(swapExact1For0(2000, wallet.address));
            expect((await pool.globalState()).price).to.not.eq(startingPrice);
            expect((await pool.globalState()).tick).to.eq(startingTick);
          });

          it('first swap in block moves tick, no initialized crossings', async () => {
            await snapshotGasCost(swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick + 1);
          });

          it('second swap in block with no tick movement', async () => {
            await swapExact1For0(expandTo18Decimals(1) / 10000n, wallet.address);
            expect((await pool.globalState()).tick).to.eq(startingTick + 1);
            await snapshotGasCost(swapExact1For0(2000, wallet.address));
            expect((await pool.globalState()).tick).to.eq(startingTick + 1);
          });
        });
      });
    });
  }

  describe('Positions', function () {
    beforeEach('load inner fixture', async () => {
      ({ advanceTime, swapExact0For1, swapExact1For0, pool, plugin, virtualPoolMock, mockPluginFactory, mint, swapToHigherPrice } = await loadFixture(
        gasTestFixture
      ));
    });

    describe('#mint', () => {
      for (const { description, bottomTick, topTick } of [
        {
          description: 'around current price',
          bottomTick: startingTick - tickSpacing,
          topTick: startingTick + tickSpacing,
        },
        {
          description: 'below current price',
          bottomTick: startingTick - 2 * tickSpacing,
          topTick: startingTick - tickSpacing,
        },
        {
          description: 'above current price',
          bottomTick: startingTick + tickSpacing,
          topTick: startingTick + 2 * tickSpacing,
        },
      ]) {
        describe(description, () => {
          it('new position mint first in range', async () => {
            await advanceTime(15);
            await snapshotGasCost(mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1)));
          });
          it('add to position existing', async () => {
            await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
            await advanceTime(15);
            await snapshotGasCost(mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1)));
          });
          it('second position in same range', async () => {
            await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
            await advanceTime(15);
            await snapshotGasCost(mint(other.address, bottomTick, topTick, expandTo18Decimals(1)));
          });
        });
      }
    });

    describe('#burn', () => {
      for (const { description, bottomTick, topTick } of [
        {
          description: 'around current price',
          bottomTick: startingTick - tickSpacing,
          topTick: startingTick + tickSpacing,
        },
        {
          description: 'below current price',
          bottomTick: startingTick - 2 * tickSpacing,
          topTick: startingTick - tickSpacing,
        },
        {
          description: 'above current price',
          bottomTick: startingTick + tickSpacing,
          topTick: startingTick + 2 * tickSpacing,
        },
      ]) {
        describe(description, () => {
          const liquidityAmount = expandTo18Decimals(1);
          beforeEach('mint a position', async () => {
            await mint(wallet.address, bottomTick, topTick, liquidityAmount);
            await advanceTime(15);
          });

          it('burn when only position using ticks', async () => {
            await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1), '0x'));
          });
          it('partial position burn', async () => {
            await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1) / 2n, '0x'));
          });
          it('entire position burn but other positions are using the ticks', async () => {
            await mint(other.address, bottomTick, topTick, expandTo18Decimals(1));
            await advanceTime(15);
            await snapshotGasCost(pool.burn(bottomTick, topTick, expandTo18Decimals(1), '0x'));
          });
        });
      }
    });

    describe('#poke', () => {
      const bottomTick = startingTick - tickSpacing;
      const topTick = startingTick + tickSpacing;

      it('best case', async () => {
        await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
        await swapExact0For1(expandTo18Decimals(1) / 100n, wallet.address);
        await pool.burn(bottomTick, topTick, 0, '0x');
        await swapExact0For1(expandTo18Decimals(1) / 100n, wallet.address);
        await advanceTime(15);
        await snapshotGasCost(pool.burn(bottomTick, topTick, 0, '0x'));
      });
    });

    describe('#collect', () => {
      const bottomTick = startingTick - tickSpacing;
      const topTick = startingTick + tickSpacing;

      it('close to worst case', async () => {
        await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
        await swapExact0For1(expandTo18Decimals(1) / 100n, wallet.address);
        await pool.burn(bottomTick, topTick, 0, '0x'); // poke to accumulate fees
        await advanceTime(15);
        await snapshotGasCost(pool.collect(wallet.address, bottomTick, topTick, MaxUint128, MaxUint128));
      });

      it('close to worst case, two tokens', async () => {
        await mint(wallet.address, bottomTick, topTick, expandTo18Decimals(1));
        await swapExact0For1(expandTo18Decimals(1) / 100n, wallet.address);
        await swapExact1For0(expandTo18Decimals(1) / 100n, wallet.address);
        await pool.burn(bottomTick, topTick, 0, '0x'); // poke to accumulate fees
        await advanceTime(15);
        await snapshotGasCost(pool.collect(wallet.address, bottomTick, topTick, MaxUint128, MaxUint128));
      });
    });
  });

  describe('Filled VolatilityOracle', function () {
    this.timeout(600_000);

    const filledStorageFixture = async () => {
      const fix = await gasTestFixture();
      await fix.advanceTime(15);
      await fix.mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
      await fix.mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
      await fix.swapExact0For1(1, wallet.address);
      expect((await fix.pool.globalState()).tick).to.eq(startingTick);

      const BATCH_SIZE = 300;
      let summaryTimeDelta = 0;
      for (let i = 0; i < 1500; i += BATCH_SIZE) {
        const batch = [];
        for (let j = 0; j < BATCH_SIZE; j++) {
          const timeDelta = (i + j) % 2 == 0 ? 60 : 90;
          summaryTimeDelta += timeDelta;
          batch.push({
            advanceTimeBy: timeDelta,
            tick: startingTick + i - j,
          });
        }
        await fix.plugin.batchUpdate(batch);
      }
      await fix.pool.advanceTime(summaryTimeDelta);

      return fix;
    };

    beforeEach('load inner fixture', async () => {
      ({ advanceTime, swapExact0For1, swapExact1For0, pool, plugin, virtualPoolMock, mockPluginFactory, mint, swapToHigherPrice } = await loadFixture(
        filledStorageFixture
      ));
    });

    describe('swaps', async () => {
      for (const isDynamicFee of [true, false]) {
        describe(isDynamicFee ? 'dynamic fee' : 'static fee', async () => {
          beforeEach(async () => {
            if (!isDynamicFee) {
              await plugin.changeFeeConfiguration({
                alpha1: 0,
                alpha2: 0,
                beta1: 0,
                beta2: 0,
                gamma1: 1,
                gamma2: 1,
                baseFee: 100,
              });
              await advanceTime(15);
              await swapExact0For1(1000, wallet.address);
            }
          });

          it('small swap with filled volatilityOracle', async () => {
            await advanceTime(15);
            await snapshotGasCost(swapExact0For1(1000, wallet.address));
          });

          it('small swap with filled volatilityOracle after 4h', async () => {
            await advanceTime(4 * 60 * 60);
            await snapshotGasCost(swapExact0For1(1000, wallet.address));
          });

          it('small swap with filled volatilityOracle after 8h', async () => {
            await advanceTime(8 * 60 * 60);
            await snapshotGasCost(swapExact0For1(1000, wallet.address));
          });

          it('small swap with filled volatilityOracle after 24h', async () => {
            await advanceTime(24 * 60 * 60 - 1);
            await swapExact0For1(1, wallet.address);
            await advanceTime(5);
            await snapshotGasCost(swapExact0For1(1000, wallet.address));
          });

          it('large swap crossing several initialized ticks', async () => {
            await advanceTime(15);
            await mint(wallet.address, startingTick - 3 * tickSpacing, startingTick - tickSpacing, expandTo18Decimals(1));
            await mint(wallet.address, startingTick - 4 * tickSpacing, startingTick - 2 * tickSpacing, expandTo18Decimals(1));
            await swapExact0For1(2, wallet.address);
            await advanceTime(15);
            await snapshotGasCost(swapExact0For1(expandTo18Decimals(1), wallet.address));
            expect((await pool.globalState()).tick).to.be.lt(startingTick - 4 * tickSpacing);
          });
        });
      }
    });
  });
});
