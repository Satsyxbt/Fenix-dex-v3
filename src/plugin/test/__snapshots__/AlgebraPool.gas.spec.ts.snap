// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee large swap crossing several initialized ticks 1`] = `207263`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle 1`] = `159489`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle after 4h 1`] = `194927`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle after 8h 1`] = `187713`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps dynamic fee small swap with filled volatilityOracle after 24h 1`] = `155978`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee large swap crossing several initialized ticks 1`] = `204377`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle 1`] = `154332`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle after 4h 1`] = `187797`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle after 8h 1`] = `198731`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Filled VolatilityOracle swaps static fee small swap with filled volatilityOracle after 24h 1`] = `153804`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn above current price burn when only position using ticks 1`] = `115316`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn above current price entire position burn but other positions are using the ticks 1`] = `108670`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn above current price partial position burn 1`] = `113470`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn around current price burn when only position using ticks 1`] = `125195`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn around current price entire position burn but other positions are using the ticks 1`] = `113078`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn around current price partial position burn 1`] = `117878`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn below current price burn when only position using ticks 1`] = `124752`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn below current price entire position burn but other positions are using the ticks 1`] = `109332`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #burn below current price partial position burn 1`] = `114132`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #collect close to worst case 1`] = `52499`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #collect close to worst case, two tokens 1`] = `70218`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint above current price add to position existing 1`] = `126222`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint above current price new position mint first in range 1`] = `280085`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint above current price second position in same range 1`] = `143322`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint around current price add to position existing 1`] = `151291`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint around current price new position mint first in range 1`] = `358199`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint around current price second position in same range 1`] = `168391`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint below current price add to position existing 1`] = `126791`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint below current price new position mint first in range 1`] = `354431`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #mint below current price second position in same range 1`] = `143891`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] Positions #poke best case 1`] = `61524`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `148898`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block with no tick movement 1`] = `148867`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block with no tick movement, static fee 1`] = `149594`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `165406`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `198800`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `148965`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `198800`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `218000`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `113409`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block with no tick movement 1`] = `113373`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `129102`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `163317`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 several large swaps with pauses 1`] = `225763`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact0For1 small swap after several large swaps with pauses 1`] = `156499`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact1For0 first swap in block moves tick, no initialized crossings 1`] = `148959`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact1For0 first swap in block with no tick movement 1`] = `148907`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap #swapExact1For0 second swap in block with no tick movement 1`] = `113434`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap farming connected first swap in block moves tick, no initialized crossings 1`] = `180120`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap farming connected first swap in block with no tick movement 1`] = `180068`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is off #swap farming connected second swap in block with no tick movement 1`] = `127495`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block moves tick, no initialized crossings 1`] = `156721`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block with no tick movement 1`] = `156690`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block with no tick movement, static fee 1`] = `149831`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block, large swap crossing a single initialized tick 1`] = `173466`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block, large swap crossing several initialized ticks 1`] = `207571`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 first swap in block, large swap, no initialized crossings 1`] = `156788`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 large swap crossing several initialized ticks after some time passes 1`] = `207571`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 large swap crossing several initialized ticks second time after some time passes 1`] = `226771`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block moves tick, no initialized crossings 1`] = `121232`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block with no tick movement 1`] = `121196`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block, large swap crossing a single initialized tick 1`] = `137162`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 second swap in block, large swap crossing several initialized ticks 1`] = `172088`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 several large swaps with pauses 1`] = `234534`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact0For1 small swap after several large swaps with pauses 1`] = `156736`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact1For0 first swap in block moves tick, no initialized crossings 1`] = `156782`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact1For0 first swap in block with no tick movement 1`] = `156730`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap #swapExact1For0 second swap in block with no tick movement 1`] = `121257`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap farming connected first swap in block moves tick, no initialized crossings 1`] = `187943`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap farming connected first swap in block with no tick movement 1`] = `187891`;

exports[`AlgebraPool gas tests [ @skip-on-coverage ] fee is on #swap farming connected second swap in block with no tick movement 1`] = `135318`;
